### PR TITLE
[Impeller] Coerce opaque ColorSourceContents to Source

### PR DESCRIPTION
**Pull request by bdero**
This coercion happens when Entities are appended to a pass, which is the moment where we will need to perform depth sorting.

### CODE REVIEW

File: impeller/entity/entity_pass.cc [hunk 1]
```diff
@@ -48,6 +48,11 @@ void EntityPass::SetDelegate(std::unique_ptr<EntityPassDelegate> delegate) {
 }
 
 void EntityPass::AddEntity(Entity entity) {
+  if (entity.GetBlendMode() == BlendMode::kSourceOver &&
+      entity.GetContents()->IsOpaque()) {
+    entity.SetBlendMode(BlendMode::kSource);
+  }
```
Comment (jonahwilliams) [inline impeller/entity/entity_pass.cc:54]: Does this happen before or after we do the opacity peephole check?

File: impeller/aiks/aiks_unittests.cc [hunk 2]
```diff
@@ -2136,5 +2136,32 @@ TEST_P(AiksTest, CanRenderOffscreenCheckerboard) {
...
+                        .color = Color::CornflowerBlue(),
+                        .blend_mode = BlendMode::kSourceOver,
+                    });
+  Picture picture = canvas.EndRecordingAsPicture();
+
+  // Extract the SolidColorSource.
+  Entity entity;
+  std::shared_ptr<SolidColorContents> contents;
+  picture.pass->IterateAllEntities([&e = entity, &contents](Entity& entity) {
+    if (ScalarNearlyEqual(entity.GetTransformation().GetScale().x, 1.618f)) {
```
Comment (bdero) [inline impeller/aiks/aiks_unittests.cc:2153]: I've been occasionally using this pattern as a sort of poor man's predicate test for inspecting what Aiks + EntityPass is doing, but I don't think it's a sustainable solution in the longterm. @chinmaygarde recall you writing a doc around this kind of thing many moons ago, but I can't seem to find it...

File: impeller/entity/entity_pass.cc (hunk 1 above)
Comment (bdero) [reply impeller/entity/entity_pass.cc:54]: This happens before (when Aiks pushes an Entity into the pass).

Hmm, it looks like we weren't handling blend modes in the opacity peephole. I added some restrictions and made it so it's able to unfold things with the Source blend mode. There's a new utility where the `Entity` handles `SetInheritedOpacity` with the blend mode in mind. Working on test cases for this at the moment.

File: impeller/aiks/paint_pass_delegate.cc [hunk 3]
```diff
@@ -72,7 +73,8 @@ bool OpacityPeepholePassDelegate::CanElide() {
 // |EntityPassDelgate|
 bool OpacityPeepholePassDelegate::CanCollapseIntoParentPass(
     EntityPass* entity_pass) {
-  if (paint_.color.alpha <= 0.0 || paint_.color.alpha >= 1.0 ||
+  if (paint_.blend_mode != BlendMode::kSourceOver ||
```
Comment (jonahwilliams) [inline impeller/aiks/paint_pass_delegate.cc:76]: This is handled in https://github.com/flutter/engine/blob/main/impeller/aiks/canvas.cc#L390

File: impeller/entity/entity_pass.cc (hunk 1 above)
Comment (bdero) [reply impeller/entity/entity_pass.cc:54]: Like if the pass has kLuminosity as its blend mode and the child entities have kHue, it'll incorrectly collapse the pass.

Comment (bdero) [reply impeller/entity/entity_pass.cc:54]: Ah nvm about the pass. I think the entities are still not being checked right, though.

**Comment by jonahwilliams** [comment]
Seems reasonable to me, though I suppose its inevitable we're going to have to be careful about the ordering of these various optimizations. It would be good to note for future that we can only re-order the entities after doing the opacity peephole as that can change the opacity and blend mode too

File: impeller/aiks/paint_pass_delegate.cc (hunk 3 above)
Comment (bdero) [reply impeller/aiks/paint_pass_delegate.cc]: Alrighty, removed this and threw in a comment.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by bdero** [comment]
My initial plan is to just have all EntityPass [Elements](https://github.com/flutter/engine/blob/fbef3281e0041412c4b5abd29682f8a137a6cf14/impeller/entity/entity_pass.h#L28-L37) treated as default/translucent and defer solving this problem.

I think we'll want to come up with a longer term plan here for pass collapsing that involves some precomputation. Maybe wrangle prioritizing other easy collapsible situations if we find they're common enough, too. Like collapsing if the pass is SourceOver+Opaque and all of its children are SourceOver/Source.

**Pull request merged**